Algorithm/QCAlgorithm.Trading.cs

```diff
@@ -604,36 +607,38 @@ public int CalculateOrderQuantity(Symbol symbol, decimal target)
             // determine the unit price in terms of the account currency
             var unitPrice = new MarketOrder(symbol, 1, UtcTime).GetValue(security);
 
-            // define lower and upper thresholds for the iteration
-            var lowerThreshold = targetOrderValue - unitPrice / 2;
-            var upperThreshold = targetOrderValue + unitPrice / 2;
+            // calculate the total margin available
+            var marginRemaining = Portfolio.GetMarginRemaining(symbol, direction);
+            if (marginRemaining <= 0) return 0;
 
-            // continue iterating while  we're still not within the specified thresholds
-            var iterations = 0;
+            // continue iterating while we do not have enough margin for the order
+            decimal marginRequired;
             var orderQuantity = 0;
-            decimal orderValue = 0;
-            while ((orderValue < lowerThreshold || orderValue > upperThreshold) && iterations < 10)
-            {
-                // find delta from where we are to where we want to be
-                var delta = targetOrderValue - orderValue;
-                // use delta value to compute a change in quantity required
-                var deltaQuantity = (int)(delta / unitPrice);
+            decimal orderValue;
+            decimal orderFees;
 
-                orderQuantity += deltaQuantity;
+            do
+            {
+                if (orderQuantity == 0)
+                {
+                    // compute the initial order quantity
+                    orderQuantity = (int)(targetOrderValue / unitPrice);
```

Comment:
I'll move this out of the loop, although it would execute the division only the first time